src/Facets/ThorSwapFacet.sol

```diff
@@ -0,0 +1,247 @@
+import { Validatable } from "../Helpers/Validatable.sol";
+import { LibSwap } from "../Libraries/LibSwap.sol";
+import { console } from "../../test/solidity/utils/Console.sol";
+
+/// @title ThorSwap Facet
+/// @author Li.Finance (https://li.finance)
+/// @notice Provides functionality for bridging through ThorSwap
+contract ThorSwapFacet is ILiFi, ReentrancyGuard, SwapperV2, Validatable {
+    /// Storage ///
+
+    bytes32 internal constant NAMESPACE =
+        keccak256("com.lifi.facets.thorswap");
+
+    address public immutable tsTokenProxy;
+
+    /// Types ///
+
+    struct Storage {
+        IThorSwap[] allowedTSRouters;
+        bool initialized;
+    }
+
+    enum RouterType {
+        Uniswap,
+        Generic,
+        Thorchain
+    }
+
+    /// @notice The struct for the ThorSwap data.
+    /// @param routerType The type of router to use
+    /// @param tsRouter The ThorSwap router
+    /// @param tcRouter The ThorChain router
+    /// @param tcVault The ThorChain vault address
+    /// @param tcMemo The ThorChain memo
+    /// @param token The token address
+    /// @param router The router address for (generic aggregators like 1inch)
+    /// @param data The data for the router
+    /// @param deadline The deadline for the swap
+    struct ThorSwapData {
+        RouterType routerType;
+        address tsRouter;
+        address tcRouter;
+        address tcVault;
+        string tcMemo;
+        address token;
+        address router;
+        bytes data;
+        uint256 deadline;
+    }
+
+    /// Errors ///
+
+    error RouterNotAllowed();
+
+    /// Events ///
+
+    event ThorSwapInitialized(IThorSwap[] allowedTSRouters);
+
+    /// @notice Initializes the ThorSwap contract
+    constructor(address _tsTokenProxy) {
+        tsTokenProxy = _tsTokenProxy;
+    }
+
+    // Init ///
+
+    /// @notice Initialize local variables for the ThorSwap Facet
+    /// @param _allowedTSRouters Allowed ThorSwap routers
+    function initThorSwap(IThorSwap[] calldata _allowedTSRouters) external {
+        LibDiamond.enforceIsContractOwner();
+
+        Storage storage s = getStorage();
+
+        if (s.initialized) {
+            revert AlreadyInitialized();
+        }
+
+        s.allowedTSRouters = _allowedTSRouters;
+        s.initialized = true;
+
+        emit ThorSwapInitialized(_allowedTSRouters);
+    }
+
+    /// @notice Bridge tokens to another chain via ThorSwap
+    /// @param _bridgeData The bridge data struct
+    /// @param _thorSwapData The ThorSwap data struct
+    function startBridgeTokensViaThorSwap(
+        ILiFi.BridgeData memory _bridgeData,
+        ThorSwapData calldata _thorSwapData
+    )
+        external
+        payable
+        nonReentrant
+        refundExcessNative(payable(msg.sender))
+        validateBridgeData(_bridgeData)
+        doesNotContainSourceSwaps(_bridgeData)
+        doesNotContainDestinationCalls(_bridgeData)
+    {
```

Comment:
We should discuss if it makes sense to validate the `destinationChainId` using a mapping to the the chain names used in the `tcMemo`.